### PR TITLE
[fx] Preserve Fx graph node order in partitioner across runs

### PR DESCRIPTION
**Pull request by kareemshaik80**
Fixes #ISSUE_NUMBER
partitioner generates different graph in recompilation on each run

### CODE REVIEW

**Comment by soulitzer** [comment]
Can you file an issue with more context (e.g. example of current behavior, expected behavior) so this can be discussed. Thanks!

**Comment by voznesenskym** [comment]
LGTM, but add a test please.

**Comment by kareemshaik80** [comment]
Thanks, Will upload test case.

**Comment by kareemshaik80** [comment]
@voznesenskym pls review uploaded test.

**Comment by ezyang** [comment]
Can you please replace the list with a dict (whose keys are the set items?) By replacing the set with a list you've made a previously O(1) test O(N), which means quadratic behavior is possible.

**Comment by kareemshaik80** [comment]
With dictionary also we will have same issue because update merge node has to still check for any duplicate keys(nodes) in the Dict this is also O(N) .

File: torch/fx/passes/infra/partitioner.py [hunk 1]
```diff
@@ -67,7 +68,7 @@ def propose_partitions(self) -> List[Partition]:
         def maybe_merge_partition(self_id: int, other_id: int):
             # merged_nodes is the union of nodes in two partition to-be-merged
             merged_nodes = copy(partitions_by_id[self_id].nodes)
-            merged_nodes.update(partitions_by_id[other_id].nodes)
+            merged_nodes.extend(n for n in partitions_by_id[other_id].nodes if n not in merged_nodes)
```
Comment (ezyang) [inline torch/fx/passes/infra/partitioner.py:71]: Because merged_nodes here is a list, the test `if n not in merged_modes` is O(n), making it quadratic.

Comment (kareemshaik80) [reply torch/fx/passes/infra/partitioner.py:71]: Okay will upload changes.

File: torch/fx/passes/infra/partitioner.py [hunk 2]
```diff
@@ -265,12 +265,13 @@ def is_transparent_output_node(node: Node, partition: Set[Node], removed_nodes:
             remove_node: Set[Node] = set()
             for node in partition.nodes:
                 if is_non_compute_node(node) and \
-                    (is_transparent_input_node(node, partition.nodes, remove_node) or
-                     is_transparent_output_node(node, partition.nodes, remove_node)):
+                    (is_transparent_input_node(node, set(partition.nodes), remove_node) or
```
Comment (ezyang) [inline torch/fx/passes/infra/partitioner.py:268]: meh, can you update these functions too

**Review by ezyang — DISMISSED** [review]

**Comment by kareemshaik80** [comment]
@ezyang @soulitzer @voznesenskym  Merge has failed due to  test_xla failure "test_resnet18 (__main__.DynamoTrainingOptimizerTest)". But xla test is not part of this repo, can someone help resolve the failure. 

I think is the following assert needs to be fixed, since preserving the order is actually reducing number of compiles.

Failure: 
2024-03-01T16:26:24.4493166Z     self.assertEqual(met.metric_data('CompileTime')[0], 7)
2024-03-01T16:26:24.4494297Z AssertionError: 6 != 7

**Comment by kareemshaik80** [comment]
> I'll do this update

Thanks @ezyang

**Comment by izaitsevfb** [comment]
@kareemshaik80 , apologies for the churn, please rebase and reland your PR once #120076 is reverted.

**Review by izaitsevfb — DISMISSED** [review]

**Comment by kareemshaik80** [comment]
> @kareemshaik80 , apologies for the churn, please rebase and reland your PR once #120076 is reverted.

@izaitsevfb what was the reason for revert?

**Comment by ezyang** [comment]
new xla companion pr https://github.com/pytorch/xla/pull/6722

**Comment by huydhn** [comment]
I think the pinned commit from this PR 8224116c3810b3dfc89708eed2e5c3a5d043167a is wrong as it refers to the commit from https://github.com/pytorch/xla/pull/6722 and the branch has been deleted.  Checking out XLA is now failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/519151a062a9bd4f0d32a9c7c7eae47d7ed847b2.  

Should it be updated to https://github.com/pytorch/xla/commit/84419438ccd706c3b56830620ad979e3a357fee3 which is the XLA trunk commit?

**Comment by kareemshaik80** [comment]
> I think the pinned commit from this PR 8224116c3810b3dfc89708eed2e5c3a5d043167a is wrong as it refers to the commit from [pytorch/xla#6722](https://github.com/pytorch/xla/pull/6722) and the branch has been deleted. Checking out XLA is now failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/519151a062a9bd4f0d32a9c7c7eae47d7ed847b2.
> 
> Should it be updated to [pytorch/xla@8441943](https://github.com/pytorch/xla/commit/84419438ccd706c3b56830620ad979e3a357fee3) which is the XLA trunk commit?
> 
> Update: I'm restoring the branch [pytorch/xla#6722](https://github.com/pytorch/xla/pull/6722) for now, it should bring back the deleted commits there till the pin is updated.

why is this PR backed out?

**Comment by kareemshaik80** [comment]
@huydhn why is the PR backed out?

**Comment by kareemshaik80** [comment]
do we get info on which tests caused for back out ?

**Comment by ezyang** [comment]
I'm just going to force push

**Review by ezyang — DISMISSED** [review]

**Comment by ezyang** [comment]
@SherlockNoMad do you think you could take a look at the error message and see if it rings a bell? The internal error is

```
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/88387c9728e02384/executorch/backends/xnnpack/test/__test_xnnpack_models__/test_xnnpack_models#link-tree/executorch/exir/backend/canonical_partitioners/pattern_op_partitioner.py", line 50, in generate_partitions_from_list_of_nodes
    partition_list = capability_partitioner.propose_partitions()
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/88387c9728e02384/executorch/backends/xnnpack/test/__test_xnnpack_models__/test_xnnpack_models#link-tree/torch/fx/passes/infra/partitioner.py", line 261, in propose_partitions
    logger.debug("partition #%s: %s", id, [node.name for node in partition.nodes])
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/88387c9728e02384/executorch/backends/xnnpack/test/__test_xnnpack_models__/test_xnnpack_models#link-tree/torch/fx/passes/infra/partitioner.py", line 261, in <listcomp>
    logger.debug("partition #%s: %s", id, [node.name for node in partition.nodes])
AttributeError: 'str' object has no attribute 'name'
```

and I'm wondering how we possibly could have gotten a string into the partition nodes set.

**Comment by kareemshaik80** [comment]
Hi @ezyang this PR has been pending for quite some time now, Can you pls prioritize this PR.

**Comment by bsochack** [comment]
@ezyang should we debug this issue? If yes, how can we reproduce it?

**Comment by kareemshaik80** [comment]
@ezyang some tests have failures reported wth no space on device. can you pls check this failure.. 

ld: final link failed: No space left on device
ERROR:common:
Traceback (most recent call last):
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/codecache.py", line 1832, in run_command_and_check
    subprocess.check_call(cmd)
  File "/opt/conda/envs/py_3.10/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['ld', '-r', '-b', 'binary', '-o', '/tmp/tmpr5ijvsv0/cyhqtrv3ddwgh6mlfwejin2fangy7fg3eu3qqwhnojalxkjr73rx/cn3whjgthqj4g4i63bmsmfbf74abjuvtky4mrqge5piaqdxgflot.o', '/tmp/tmpr5ijvsv0/cyhqtrv3ddwgh6mlfwejin2fangy7fg3eu3qqwhnojalxkjr73rx/cn3whjgthqj4g4i63bmsmfbf74abjuvtky4mrqge5piaqdxgflot.bin']' returned non-zero exit status 1.

**Comment by kareemshaik80** [comment]
@ezyang regarding OOM issues if these are related to PR chagnes. Can you pls check?

torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 82.00 MiB. GPU 0 has a total capacity of 21.98 GiB of which 64.44 MiB is free. Process 21268 has 21.90 GiB memory in use. Of the allocated memory 21.13 GiB is allocated by PyTorch, and 433.16 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

The above exception was the direct cause of the following exception:

**Comment by ezyang** [comment]
To make it easier to diagnose the executorch failure, I'm recreating this PR as a phabricator diff https://www.internalfb.com/diff/D58372936

**Review by ezyang — APPROVED** [review]

**Comment by nWEIdia** [comment]
Does gluon_inception_v3 accuracy look more red after this merge?

**Pull request closed without merging**